plugins/hls-alternate-number-format-plugin/src/Ide/Plugin/Conversion.hs

```diff
@@ -0,0 +1,170 @@
+
+removeCurrentFormat :: Literal -> [FormatType] -> [FormatType]
+removeCurrentFormat lit fmts = case getSrcText lit of
+    Just src -> foldl (flip delete) fmts (sourceToFormatType src)
+    Nothing  -> fmts
+
+hexRegex, hexFloatRegex, binaryRegex, octalRegex, decimalRegex, numDecimalRegex :: Text
+hexRegex = "0[xX][a-fA-F0-9]+"
+hexFloatRegex = "0[xX][a-fA-F0-9]+\\.[a-fA-F0-9]+(p[+-]?[0-9]+)?"
+binaryRegex = "0[bB][0|1]+"
+octalRegex = "0[oO][0-8]+"
+decimalRegex = "\\d+"
+numDecimalRegex = "[0-9]+\\.[0-9]+[eE][+-]?[0-9]+"
+
+-- we export the regex for tests, but we want to be explicit in our matches
+-- so we need to match the beginning/end of the source text
+wrap :: Text -> Text
+wrap regex = "^" <> regex <> "$"
+
+sourceToFormatType :: Text -> [FormatType]
+sourceToFormatType srcText
+    | srcText =~ wrap hexRegex = [IntFormat HexFormat]
+    | srcText =~ wrap hexFloatRegex = [FracFormat HexFloatFormat]
+    | srcText =~ wrap octalRegex = [IntFormat OctalFormat]
+    | srcText =~ wrap binaryRegex = [IntFormat BinaryFormat]
+    -- can either be a NumDecimal or just a regular Fractional with an exponent
+    -- otherwise we wouldn't need to return a list
+    | srcText =~ wrap numDecimalRegex  = [IntFormat NumDecimalFormat, FracFormat ExponentFormat]
+    -- just assume we are in base 10 with no decimals
+    | otherwise = [AnyFormat DecimalFormat]
+
+-- grab the Numeric related extensions that are turned on for a file
+toFormatTypes :: [Extension] -> [FormatType]
+toFormatTypes =  (<>) [IntFormat HexFormat, IntFormat OctalFormat, FracFormat ExponentFormat, AnyFormat DecimalFormat]
+                 . mapMaybe (`lookup` numericPairs)
+
+-- current list of Numeric related extensions
+-- LexicalNegation --- 9.0.1 > --- superset of NegativeLiterals
+numericPairs :: [(Extension, FormatType)]
+numericPairs = [(NumericUnderscores, NoFormat), (NegativeLiterals, NoFormat)] <> intPairs <> fracPairs
+
+numericFormats :: [FormatType]
+numericFormats = map snd numericPairs
+
+intPairs :: [(Extension, FormatType)]
+intPairs = [(BinaryLiterals, IntFormat BinaryFormat), (NumDecimals, IntFormat NumDecimalFormat)]
+
+intFormats :: [FormatType]
+intFormats = map snd intPairs
+
+fracPairs :: [(Extension, FormatType)]
+fracPairs = [(HexFloatLiterals, FracFormat HexFloatFormat)]
+
+fracFormats :: [FormatType]
+fracFormats = map snd fracPairs
+
+-- Generate up to 3 possible choices where:
+-- dropWhile (\d -> val `div` d) > 1000) implies we want at MOST 3 digits to left of decimal
+-- takeWhile (val >) implies we want to stop once we start to get numbers like: 0.1e[N]
+-- take 3 implies we want at most three choices which will center around the format:
+--    - 500.123e4
+--    - 50.0123e5
+--    - 5e.00123e6
+-- NOTE: showEFloat would also work, but results in only one option
+toNumDecimal :: Integer -> [Text]
+toNumDecimal val = map (showNumDecimal val) $ take 3 $ takeWhile (val >= ) $ dropWhile (\d -> (val `div` d) > 1000) divisors
```

Comment:
Given the comment above, this feels like a slightly inaccurate name for the function (sorry, I know, I am nitpicky)